lib/std/os/linux.zig

```diff
@@ -121,6 +121,16 @@ pub fn utimensat(dirfd: i32, path: ?[*:0]const u8, times: *const [2]timespec, fl
     return syscall4(.utimensat, @bitCast(usize, @as(isize, dirfd)), @ptrToInt(path), @ptrToInt(times), flags);
 }
 
+pub fn fallocate(fd: i32, mode: i32, offset: usize, len: usize) usize {
```

Comment:
should `offset` be a `u64`?

Reply:
> should offset be a u64?

Both `offset` and `usize` are `off_t` (so `u64`) and may need special care when passed to the kernel and be split in `u32`/`u32`.

Reply:
> Both offset and usize are off_t (so u64)

I assumed that `off_t` is guaranteed to be at least an `int` but is not always `u64`, which is why I went with `usize`, thinking that would be 32-bit on 32-bit machines else 64-bit. You will know better than me!

`off_t` seems complicated, according to https://stackoverflow.com/a/14351239/2138743:

> This is a signed integer type used to represent file sizes. 
> In the GNU C Library, this type is no narrower than int.
> If the source is compiled with _FILE_OFFSET_BITS == 64 this 
> type is transparently replaced by off64_t.

How would we know if the source is compiled with `_FILE_OFFSET_BITS == 64`?

Reply:
> may need special care when passed to the kernel and be split in u32/u32.

What would this special care look like?

Reply:
> I assumed that off_t is guaranteed to be at least an int but is not always u64, which is why I went with usize, thinking that would be 32-bit on 32-bit machines else 64-bit. You will know better than me!

Don't mix up libc definitions and kernel ones, the latter is defined as `long long`.

> How would we know if the source is compiled with _FILE_OFFSET_BITS == 64?

Most of the stdlib assumes that's true, use `off_t` in the libc definition (even though that's signed...).

> What would this special care look like in more detail? I'm new to this, so if you could provide a sample that would be a huge help.

32bit kernels work on 32bit registers so passing the offset requires some care. Check out musl's `fallocate.c` wrapper for more info, I usually double-check if the values are sane using `strace`.

> What signature would you propose we use here?

Follow the libc signature, they have to match:
```zig
pub extern "c" fn fallocate(fd: fd_t, mode: c_int, offset: off_t, len: off_t) c_int
```

> I assume these would go in os/bits/linux.zig?

Yup

Reply:
> With that in mind, would it be fine if we do a branch on @sizeOf(usize) and then select syscall4 (64-bit kernel) or syscall6 (32-bit kernel)?

Yep

> What would be the ziggest way of splitting the long long into two longs

Shift&truncate

Reply:
So musl's `fallocate.c` looks like this:

```c
int fallocate(int fd, int mode, off_t base, off_t len)
{
    return syscall(SYS_fallocate, fd, mode, __SYSCALL_LL_E(base),
        __SYSCALL_LL_E(len));
}
```

Where `__SYSCALL_LL_E` is defined as an identity function to pass one `long long` parameter, or else two `long` parameters to the variadic `syscall` function, depending on the architecture:

```c
#define __SYSCALL_LL_E(x) (x)
```

or

```c
#define __SYSCALL_LL_E(x) \
((union { long long ll; long l[2]; }){ .ll = x }).l[0], \
((union { long long ll; long l[2]; }){ .ll = x }).l[1]
```

musl defines these per architecture, using a single `long long` parameter for:

```
aarch64
mips64
powerpc64
riscv64
s390x
x86_64
```

and two `long` parameters for:

```
arm
i386
mips
powerpc
```

With that in mind, would it be fine if we do a branch on `@sizeOf(usize)` and then select `syscall4` (64-bit kernel) or `syscall6` (32-bit kernel)?

What would be the *ziggest* way of splitting the `long long` into two `longs`? `@bitCast([2]u32, offset)` assuming `offset` becomes a `u64` in our `fallocate()` function signature as @daurnimator suggested?

Reply:
Done

Reply:
...I see `llseek()` simply does `@truncate(usize, offset >> 32), @truncate(usize, offset)`